src/streamlink/utils/formatter.py

```diff
@@ -0,0 +1,64 @@
+import re
+
+from streamlink_cli.compat import is_win32
+from streamlink_cli.constants import DEFAULT_STREAM_METADATA, FS_SAFE_REPLACEMENT_CHAR, INVALID_FILENAME_CHARS
+
+
+class Formatter(dict):
+    def __init__(self, plugin, args):
+        self.plugin = plugin
+        self.args = args
+        self.cache = self.Cache(self.plugin)
+
+    def __missing__(self, k):
+        if k in ("title", "author", "category", "game", "url"):
+            self[k] = self._make_fs_safe(self.cache[k])
+        else:
+            return f"{{{k}}}"
+
+        return self[k]
+
+    def _make_fs_safe(self, unsafe_str):
+        if self.args.fs_safe_rules in INVALID_FILENAME_CHARS:
+            rules = self.args.fs_safe_rules
+        else:
+            if is_win32:
+                rules = "Windows"
+            else:
+                rules = "POSIX"
+
+        sub_re = re.compile(
+            "[{:s}]".format(re.escape(INVALID_FILENAME_CHARS[rules])), re.UNICODE
+        )
+
+        return sub_re.sub(FS_SAFE_REPLACEMENT_CHAR, unsafe_str)
+
+    def get_formatted_filename(self, unformatted_filename):
+        return unformatted_filename.format_map(self)
+
+    def get_formatted_title(self, unformatted_title):
+        if self.args.title:
+            return unformatted_title.format_map(self.cache)
+        else:
+            return self.args.url
+
+    class Cache(dict):
+        def __init__(self, plugin):
+            self.plugin = plugin
+
+        def __missing__(self, k):
+            if k == "title":
+                v = self.plugin.get_title() or DEFAULT_STREAM_METADATA["title"]
```

Comment:
all `DEFAULT_STREAM_METADATA` are pretty stupid, when it get used for a filename
for a player title it is like whatever, but when used for a file it will try  to create files with the same name.

https://github.com/streamlink/streamlink/blob/6cdc3ebb141f83c8fe34f2426685443fa1e1b7de/src/streamlink_cli/constants.py#L11-L16

Reply:
I don't really see how it's that much of a problem using these default strings in filenames when title/author/category/game cannot be retrieved via the plugin.  They seem as good as any to me.

Potentially, yes, if a plugin does not implement any of the metadata methods then it's going to fall back to a filename that will possibly be a duplicate if Streamlink has been called before with a plugin where this has also happened, but the file will not be overwritten unless `--force` has been specified.  I don't think this code should have to do anything to avoid that situation.  Why should it need to do so?